File: baekjoon/_4949.py
sentenses = []

while True:
    sentense = input()
    if sentense == '.':
        break
    sentenses.append(sentense)

for sentense in sentenses:
    flag = False
    stack = []
    for word in sentense:
        if (word == '(' or word == '['):
            stack.append(word)

        elif (word == ')'):
            if (not stack or stack[-1] != '('):
                flag = True
                break
            else:
                stack.pop()

        elif (word == ']'):
            if (not stack or stack[-1] != '['):
                flag = True
                break
            else:
                stack.pop()
        else:
            continue

    if (not stack and flag == False):
        print('yes')
    else:
        print('no')

File: baekjoon/_4949.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    while (true) {
        string s;
        getline(cin, s);

        stack<char> S;
        bool flag = true;

        if (s == ".") {
            break;
        }

        for (char c : s) {
            if (c == '(' || c == '[') {
                S.push(c);
            } else if (c == ')') {
                if (S.empty() || S.top() != '(') {
                    flag = false;
                    break;
                }
                S.pop();
            } else if (c == ']') {
                if (S.empty() || S.top() != '[') {
                    flag = false;
                    break;
                }
                S.pop();
            }
        }

        if (flag && S.empty()) {
            cout << "yes" << "\n";
        } else {
            cout << "no" << "\n";
        }
    }
}